### ISSUE TITLE
Wrong location for error with polymorphic variant patterns

### ISSUE THREAD

**Issue by vicuna** [ocaml/ocaml#6330]
**Original bug ID:** 6330
**Reporter:** @lpw25
**Assigned to:** @garrigue
**Status:** acknowledged (set by @gasche on 2014-02-23T18:02:16Z)
**Resolution:** open
**Priority:** normal
**Severity:** minor
**Version:** 4.01.0
**Target version:** later
**Category:** typing
**Monitored by:** @yallop @hcarty

## Bug description

The following error should really be located at the second case:

```
   # let f (x: [< `Foo of int | `Bar of float]) = 
       match x with
        `Foo x -> x
      | `Baz -> 1
      | _ -> 0;;
   Characters 65-71:
        `Foo x -> x
        ^^^^^^
    Error: This pattern matches values of type [> `Baz | `Foo of 'a ]
           but a pattern was expected which matches values of type
           [< `Bar of float | `Foo of int ]
           The second variant type does not allow tag(s) `Baz
```

**Comment by vicuna** [comment]
**Comment author:** @gasche

I did a bit of work to understand where the issue comes from -- and found out that it was not easily fixable. This will probably be obvious to Leo and Jacques, but I'll still summarize the issue in case anyone is interested.

The error location comes from the way pattern-matching of polymorphic variants are type-checked. First, the type of all the branches are unified together -- and passed through a bit of variant-pressuring magic I don't understand. Then, they are unified with the type of the pattern scrutinee. In this example, unifying both patterns together gives them *both* a type of the form ``[> `Foo of '_a | `Baz ]``. When the `` `Foo`` pattern is then unified with the type ``[< `Foo of int | `Bar of float]``, you get an error (located at `` `Foo``) about the type `` `Baz`` not being possible.

For someone like me that is not familiar with type inference of polymorphic variants, it's not obvious why it wouldn't work to first unify each branch with (an instance of) the scrutinee's type, and then unify the branch together -- a clash between one of the branch and the scrutinee would then be reported where expected. However, doing this does not work (and the implementation is very clear about the fact that the existing order is intended) on some other examples; for example,

    fun x -> match (x : [< `A ]) with `A | `B -> ();;

is accepted by the current type-checking technique (with a warning about the `` `B`` being dead code), but would be rejected by trying to unify `` `B``'s type with ``[< `A ]``.

TL;DR: it's not a silly typo in the location-handling code, it's something subtle about type-checking of polymorphic variants, that is probably not going to change soon - or it needs an expert.

**Comment by garrigue** [comment]
Update: if you write your code in the following way, you get the error at the right position.
```ocaml
let f (x: [`Foo of int | `Bar of float]) = match x with
          `Foo x -> x
        | `Baz -> 1
        | _ -> 0
```
So why doesn't it work for annotations of the form `[< ...]` ?
As @gasche explained, the logic used to decide whether a variant type should be closed or open requires that we do not propagate "potential presence information" to the pattern-matching typer. So this type annotation is first ignored, and then unified with each case. And we get the clash when unifying with the first case. It might be possible to make the location wider (i.e. the whole pattern-matching), but I do not see how to make it point to the second case.
Can we just decide that we won't fix this one?

**Comment by gasche** [comment]
I propose to close unless @lpw25 has a different suggestion.

**Comment by lpw25** [comment]
Seems not worth the trouble to fix
